blobfuse/src/BlobStreamer.cpp

```diff
@@ -0,0 +1,327 @@
+int BlobStreamer::ReadFile(const char* file_name, uint64_t offset, uint64_t length, char* out)
+{
+    int len = 0;
+
+    // If block caching is not allowed as per config then stream data directly from container.
+    if (max_blocks_per_file <= 0) {
+        // Get data in form of a stream and fill the output buffer with data retreived
+        std::stringstream os;
+        //syslog(LOG_DEBUG, "%s : Read (%lu, %lu)", file_name, offset, length);
+        azclient->DownloadToStream(file_name, os, offset, length);
+        if (errno != 0)
+        {
+            if (errno == 416) {
+                // Range given the request is invalid so we mark it as end of file
+                errno = 0;
+                return 0;
+            }
+
+            int storage_errno = errno;
+            syslog(LOG_ERR, "Failed to download block of %s with offset %lu.  Errno = %d.\n", file_name, offset, storage_errno);
+            return -storage_errno;
+        }
+
+        len = os.str().size();
+        os.read(out, len);
+        out[len] = '\0';
+        return len;
+    } 
+
+    // Caching of block is allowed so we need to check block exists or not
+    m_mutex.lock();
+    auto iter = file_map.find(file_name);
+    if(iter == file_map.end()) {
+        m_mutex.unlock();
+        return 0;
+    }
+    StreamObject* obj = iter->second;
+    m_mutex.unlock();
+
+    while (length > 0) {
+        //  At max the data requested may overlap two blocks
+        //  as soon as we get the full data we return back
+        BlobBlock* block = GetBlock(file_name, offset, obj);
+        if (block == NULL) {
+            if (errno == 416) {
+                // Range given the request is invalid so we mark it as end of file
+                errno = 0;
+                return 0;
+            }
+            // For some reason we failed to get the block object
+            syslog(LOG_ERR, "Failed to get block for %s with offset %lu", file_name, offset);
+            return -errno;
+        }
+
+        // Based on offset and block being used calculate the start offset inside the block
+        uint64_t start_offset = offset - block->start_offset;
+        uint64_t pending_data = block->data_length - start_offset;
+
+        syslog(LOG_DEBUG, "%s : Read Block (%lu, %lu, %d)  Request (%lu, %lu)  Read (%lu, %lu)",
+                file_name, block->start_offset, ((block->data_length + block->start_offset) - 1), block->last, 
+                offset, length, 
+                start_offset, pending_data);
+
+        if (pending_data < length) {
+            // Either request overlaps two blocks or request exceeds the file size
+            // So read the remaining data from this block and decide
+            block->buff.seekg(start_offset, std::ios::beg);
+            block->buff.read((out + len), pending_data);
+            block->lck.unlock();
+
+            len += pending_data;
+
+            if (block->last) {
+                // This block is the last so terminate the loop
```

Comment:
I don't think we will get a case where we hit this - since the kernel knows when its the end of the file and will request the data size accordingly. I assume this is defensive programming?

Reply:
'last' flag in block just means this is the last block from file. For the concern on the kernel reading only upto size of file, I have seen cases where kernel tries to read beyond size and we need to return back 0 byte data to mark end of file. Not sure under what scenario that was hit but that was the reason to add this condition in first place.